include/small/slab.h

```diff
@@ -0,0 +1,125 @@
+#pragma once
+/*
+ * Copyright 2010-2016, Tarantool AUTHORS, please see AUTHORS file.
+ *
+ * Redistribution and use in source and binary forms, with or
+ * without modification, are permitted provided that the following
+ * conditions are met:
+ *
+ * 1. Redistributions of source code must retain the above
+ *    copyright notice, this list of conditions and the
+ *    following disclaimer.
+ *
+ * 2. Redistributions in binary form must reproduce the above
+ *    copyright notice, this list of conditions and the following
+ *    disclaimer in the documentation and/or other materials
+ *    provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY <COPYRIGHT HOLDER> ``AS IS'' AND
+ * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED
+ * TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+ * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL
+ * <COPYRIGHT HOLDER> OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT,
+ * INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+ * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR
+ * BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
+ * LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF
+ * THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
+ * SUCH DAMAGE.
+ */
+#include "util.h"
+
+/** Aligned size of slab meta. */
+static inline uint32_t
+slab_sizeof()
+{
+	return small_align(sizeof(struct slab), sizeof(intptr_t));
+}
+
+/** Useful size of a slab. */
+static inline uint32_t
+slab_capacity(struct slab *slab)
+{
+	return slab->size - slab_sizeof();
+}
+
+static inline struct slab *
+slab_from_data(void *data)
+{
+	return (struct slab *) ((char *) data - slab_sizeof());
+}
+
+static inline void *
+slab_data(struct slab *slab)
+{
+	return (char *) slab + slab_sizeof();
+}
+
+#if !defined(SLAB_MIN_ORDER0_SIZE)
+/**
+ * By default get this value from sysconf, but it should be
+ * possible to change this value from CMake for test purpose.
+ */
+#define SLAB_MIN_ORDER0_SIZE small_getpagesize()
+#endif /* !defined(SLAB_MIN_ORDER0_SIZE) */
+
+static inline void
+slab_order_create(struct slab_cache *cache, uint32_t slab_size)
+{
+	/*
+	 * We have a fixed number of orders (ORDER_MAX); calculate the size of
+	 * buddies in the smallest order, given the size of the slab size in
+	 * the slab arena.
+	 */
+	long min_order0_size = SLAB_MIN_ORDER0_SIZE;
+	assert((long)slab_size >= min_order0_size);
+	cache->order_max = small_lb(slab_size / min_order0_size);
+	if (cache->order_max > ORDER_MAX - 1)
+		cache->order_max = ORDER_MAX - 1;
+
+	cache->slab_size = slab_size;
+	cache->order0_size = slab_size >> cache->order_max;
+	cache->order0_size_lb = small_lb(cache->order0_size);
```

Comment:
This code won't compile unless this header is included after `slab_cache` is defined. Please don't rely on `include` directive order.

Reply:
Fixed as discussed f2f:
- ASAN slab cache implementation does not follow regular implementation slab order so we don't need to factor it out to reuse
- functions like `slab_sizeof` are placed at the end of `slab_cache.h` in the common section
- `slab.h` is dropped